src/dataflow/components/dataflow-program.tsx

```diff
@@ -139,46 +145,15 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
                                             units: ch.units,
                                             value: Number(ch.value)};
               this.channels.push(nci);
-              newChannel = true;
             }
-          });
-        });
-        // update any existing blocks since we found a new sensor or relay
-        this.updateChannelNodes(newChannel, newChannel);
-      });
-
-      // Can this auto-call processing and be in the sensor component instead of doing this n^2 loop
-      autorun(() => {
-        const { hubStore } = this.stores;
-        let change = false;
-        hubStore.hubs.forEach(hub => {
-          hub.hubChannels.forEach(ch => {
+            // store sensor value for channel
             const chValue = Number.parseFloat(ch.value);
-            const nci = this.channels.find(ci => ci.channelId === ch.id);
-            if (nci && Number.isFinite(chValue)) {
-              nci.value = chValue;
-            }
-            if (Number.isFinite(chValue) && ch.type !== "relay") {
-              const nodes = this.programEditor.nodes.filter((n: Node) => n.data.sensor === ch.id);
-              if (nodes) {
-                nodes.forEach((n: Node) => {
-                  const sensorSelect = n.controls.get("sensorSelect") as SensorSelectControl;
-                  if (sensorSelect && sensorSelect.getSensorValue() !== chValue) {
-                    sensorSelect.setSensorValue(chValue);
-                    change = true;
-                  }
-                });
-              }
+            const chInfo = this.channels.find(ci => ci.channelId === ch.id);
+            if (chInfo && Number.isFinite(chValue)) {
+              chInfo.value = chValue;
             }
           });
         });
-        this.updateChannelNodes(true, false);
-        if (change) {
-          (async () => {
-            await this.programEngine.abort();
-            await this.programEngine.process(this.programEditor.toJSON());
```

Comment:
This abort/process cycle now happens only in the heartbeat.  Should reduce competing calls if heartbeat and sensor update happen at same time.